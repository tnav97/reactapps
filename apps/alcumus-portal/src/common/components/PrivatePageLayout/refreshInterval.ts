const MAXIMUM_REFRESH_INTERVAL_IN_MINUTES = 15;

export function getTokenRefreshIntervalInMilliseconds(
  secondsTillExpiry: number
): number {
  // Refresh tokens before a minute of expiry
  // azuread access tokens expire in 60 minutes by default
  // keycloak access tokens expire in 5 minutes by default
  const suggestedExpiryIntervalInMinutes = secondsTillExpiry / 60 - 1;
  const tokenRefreshIntervalInMilliseconds =
    Math.min(
      MAXIMUM_REFRESH_INTERVAL_IN_MINUTES,
      suggestedExpiryIntervalInMinutes
    ) *
    60 *
    1000;
  return tokenRefreshIntervalInMilliseconds;
}
