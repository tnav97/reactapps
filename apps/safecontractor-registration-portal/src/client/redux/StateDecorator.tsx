import React from 'react';
import { Provider } from 'react-redux';
import configureStore from './configureStore';
import { PersistGate } from 'redux-persist/integration/react';

interface StateDecoratorProps {
  children?: React.ReactNode;
}

export default function StateDecorator({
  children,
}: StateDecoratorProps): JSX.Element {
  const { store, persistor } = configureStore();

  return (
    <Provider store={store}>
      <PersistGate persistor={persistor}>{children}</PersistGate>
    </Provider>
  );
}
