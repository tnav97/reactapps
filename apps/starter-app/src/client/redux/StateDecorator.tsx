import React from 'react';
import { Provider } from 'react-redux';
import configureStore from './configureStore';

interface StateDecoratorProps {
  initialState?: object;
  children?: React.ReactNode;
}

export default function StateDecorator({
  initialState,
  children,
}: StateDecoratorProps): JSX.Element {
  const store = configureStore(initialState);

  return <Provider store={store}>{children}</Provider>;
}
