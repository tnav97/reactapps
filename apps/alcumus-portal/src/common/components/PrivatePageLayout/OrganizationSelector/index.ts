import { connect } from 'react-redux';
import { RootReducerState } from '../../../../client/redux/reducers';
import { OrganizationSelector } from './OrganizationSelector';

function mapStateToProps({ currentOrganization, user }: RootReducerState) {
  return {
    user: user,
    currentOrganization: currentOrganization.currentOrganization,
  };
}

export default connect(mapStateToProps)(OrganizationSelector);
