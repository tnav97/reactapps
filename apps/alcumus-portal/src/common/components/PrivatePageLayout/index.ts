import { connect } from 'react-redux';
import { fetchUserProfile, refreshTokens } from '../../../client/redux/actions';
import { RootReducerState } from '../../../client/redux/reducers';
import PrivatePageLayout from './PrivatePageLayout';

function mapStateToProps({
  user,
  features,
  currentOrganization,
  auth,
}: RootReducerState) {
  return {
    user,
    isFetchingUserProfile: user.isFetching,
    error: user.error,
    useAzureAd: features.useAzureAd,
    isLoggedIn: !auth.loggedOut,
    currentOrganization: currentOrganization.currentOrganization,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    fetchUserProfile: () => dispatch(fetchUserProfile()),
    refreshTokens: (refreshToken: string) =>
      dispatch(refreshTokens(refreshToken)),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(PrivatePageLayout);
