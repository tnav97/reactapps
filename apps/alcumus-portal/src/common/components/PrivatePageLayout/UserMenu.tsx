import React from 'react';
import { Link } from 'react-router-dom';
import { useTranslation } from 'react-i18next';
import { createStyles, makeStyles } from '@mui/styles';
import { Image, OutlinedIcon, StyleVariables } from '@alcumus/components';
import clsx from 'clsx';
import { UserProfile } from '../../types';
import OrganizationSelector from './OrganizationSelector';
import { IconButton, Menu, MenuItem, Theme } from '@mui/material';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    userContent: {
      display: 'flex',
      justifyContent: 'flex-start',
      alignItems: 'center',
    },
    userEmail: {
      fontSize: StyleVariables.fonts.size.xs,
      fontWeight: StyleVariables.fonts.weight.bold,
      color: StyleVariables.colors.text.subdued,
    },
    menuButton: {
      display: 'flex',
      color: StyleVariables.colors.icon.default,
      justifyItems: 'center',
    },
    userProfileMenu: {
      display: 'flex',
      marginLeft: '1rem',
      alignItems: 'center',
    },
    userProfile: {
      width: '35px',
      height: '35px',
    },
    userProfileMenuDropdown: {
      '& > .MuiPaper-root.MuiMenu-paper': {
        top: '3.5rem !important',
      },
    },
    [theme.breakpoints.only('xs')]: {
      squeezeText: {
        textOverflow: 'ellipsis',
        width: '43px',
      },
    },
    desktopUserMenu: {
      zIndex: 1,
      display: 'inline-block',
      [theme.breakpoints.up('xs')]: {
        display: 'flex',
      },
      [theme.breakpoints.down('xs')]: {
        display: 'none',
      },
    },
    mobileUserMenu: {
      zIndex: 2,
      [theme.breakpoints.up('xs')]: {
        display: 'none',
      },
      [theme.breakpoints.down('xs')]: {
        display: 'flex',
      },
    },
  })
);

interface UserMenuProps {
  user?: UserProfile;
  onLogout: () => void;
}

export default function UserMenu({ onLogout, user }: UserMenuProps) {
  const classes = useStyles();
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);
  const isMenuOpen = Boolean(anchorEl);
  const { t } = useTranslation('translation');

  const handleProfileMenuOpen = (event: React.MouseEvent<HTMLElement>) => {
    setAnchorEl(event.currentTarget);
  };

  const handleMenuClose = () => {
    setAnchorEl(null);
  };

  const handleLogout = () => {
    onLogout();
    setAnchorEl(null);
  };

  return (
    <React.Fragment>
      <div className={classes.desktopUserMenu}>
        {(user?.organizations || []).length > 0 && <OrganizationSelector />}
        <div className={classes.userProfileMenu}>
          <Image
            src="/icons/Profile.svg"
            alt="User Profile"
            aria-label="account of current user"
            aria-controls="primary-search-account-menu"
            aria-haspopup="true"
            data-testid="profileImage"
            className={classes.userProfile}
            onClick={handleProfileMenuOpen}
          />
          <Menu
            anchorEl={anchorEl}
            anchorOrigin={{ vertical: 'top', horizontal: 'right' }}
            id="primary-search-account-menu"
            keepMounted
            transformOrigin={{ vertical: 'top', horizontal: 'right' }}
            open={isMenuOpen}
            onClose={handleMenuClose}
            className={classes.userProfileMenuDropdown}
          >
            <MenuItem data-testid="settings" component={Link} to="/settings">
              {t('settings')}
            </MenuItem>
            <MenuItem data-testid="logoutLink" onClick={handleLogout}>
              {t('logout')}
            </MenuItem>
          </Menu>
        </div>
      </div>
      <div className={classes.mobileUserMenu}>
        <IconButton
          color="inherit"
          aria-label="Settings"
          className={clsx(classes.menuButton)}
          data-testid="settingsIcon"
          onClick={handleProfileMenuOpen}
        >
          <OutlinedIcon icon="settings" />
        </IconButton>
      </div>
    </React.Fragment>
  );
}
