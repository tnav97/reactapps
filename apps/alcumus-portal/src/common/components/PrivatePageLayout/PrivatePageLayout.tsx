import React, { useCallback, useContext, useEffect } from 'react';
import { createStyles, makeStyles } from '@mui/styles';
import { ListItem, ListItemIcon, Theme } from '@mui/material';
import MToolbar from '@mui/material/Toolbar'
import { Redirect, useHistory, useLocation } from 'react-router-dom';
import qs from 'querystring';
import { useInterval } from '@alcumus/hooks';
import {
  Button,
  Drawer,
  DrawerContent,
  DrawerItem,
  IDrawerContext,
  Navigation,
  Page,
  RegularIcon,
  StyleVariables,
} from '@alcumus/components';
import store from 'store';
import UserMenu from './UserMenu';
import PageContext from '../../context/PageContext';
import { UserProfile } from '../../types';
import getLoginPageUrl from '../../../lib/utils/getLoginPageUrl';
import clsx from 'clsx';
import { ToastContextProvider } from '../../context/ToastContext';
import { getTokenRefreshIntervalInMilliseconds } from './refreshInterval';
import Analytics from '@alcumus/analytics-package';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      display: 'flex',
    },
    content: {
      flexGrow: 1,
      padding: theme.spacing(3),
      [theme.breakpoints.only('xs')]: {
        padding: theme.spacing(1),
      },
    },
    portalPage: {
      backgroundColor: (props: { backgroundColor?: string }) =>
        props.backgroundColor || StyleVariables.colors.background.default,
      padding: '1rem 0',
    },
    formControl: {
      minWidth: theme.spacing(30),
    },
    backButtonContainer: {
      marginTop: '36px',
      marginBottom: '-8px',
      padding: '0 24px',
    },
    backButtonContainerSmall: {
      padding: '0',
      paddingLeft: '4px',
    },
  })
);

interface PrivatePageLayoutProps {
  user?: UserProfile;
  isFetchingUserProfile?: boolean;
  error?: string;
  children: React.ReactNode;
  useAzureAd?: boolean;
  fetchUserProfile: () => void;
  refreshTokens: (refreshToken: string) => void;
  drawerItems?: Array<DrawerItem>;
  showBackButton?: boolean;
  backgroundColor?: string;
}

function BackButton() {
  const history = useHistory();
  const drawerContext = useContext<IDrawerContext>(Drawer.Context);
  const classes = useStyles({});

  return (
    <div
      className={clsx([
        classes.backButtonContainer,
        {
          [classes.backButtonContainerSmall]: !drawerContext.isOpen,
        },
      ])}
    >
      {drawerContext.isOpen ? (
        <Button
          onClick={() => history.push('/')}
          variant="outlined"
          fullWidth={drawerContext.isOpen}
          rounded
          data-testid="exit-settings-button"
        >
          <RegularIcon
            icon="west"
            style={{ fontSize: 'inherit', marginRight: '8px' }}
          />
          {drawerContext.isOpen && 'Exit settings'}
        </Button>
      ) : (
        <ListItem onClick={() => history.push('/')}>
          <ListItemIcon>
            <RegularIcon icon="west" />
          </ListItemIcon>
        </ListItem>
      )}
    </div>
  );
}

export default function PrivatePageLayout({
  user,
  isFetchingUserProfile,
  error,
  useAzureAd = false,
  fetchUserProfile,
  refreshTokens,
  children,
  drawerItems = [],
  showBackButton,
  backgroundColor,
}: PrivatePageLayoutProps) {
  const classes = useStyles({ backgroundColor });
  const history = useHistory();
  const location = useLocation();
  const { logout } = qs.parse(location.search.substring(1));
  const refreshToken = store.get('refreshToken');
  const secondsTillExpiry = Number(store.get('expiresIn'));

  // refresh tokens to continue idle session in the background
  useInterval(() => {
    if (refreshToken) {
      refreshTokens(refreshToken);
    }
  }, getTokenRefreshIntervalInMilliseconds(secondsTillExpiry));

  // fetch user profile if unavailable
  useEffect(() => {
    if (Boolean(logout) === true) {
      history.push('/logout');
    } else if (!user?.email && !isFetchingUserProfile && !error) {
      fetchUserProfile();
    }
  }, [history, fetchUserProfile]);

  useEffect(() => {
    if (user && user.firstName) {
      Analytics.getInstance().identify(`${user?.firstName} ${user.lastName}`);
    }
  }, [user]);

  // Hard redirect to login page if logged out and using azure ad
  useEffect(() => {
    if (useAzureAd && (!refreshToken || !secondsTillExpiry)) {
      window.location.href = getLoginPageUrl();
    }
  }, [useAzureAd, refreshToken, secondsTillExpiry]);

  const handleLogout = useCallback(() => {
    history.push('/logout');
  }, [history]);

  // if there are no tokens found in local storage, then redirect to /login
  if (!useAzureAd && (!refreshToken || !secondsTillExpiry)) {
    return <Redirect to={getLoginPageUrl()} />;
  }

  return (
    <Page className={classes.portalPage}>
      <PageContext.Provider value={{ user }}>
        <ToastContextProvider>
          <div className={classes.root}>
            <Navigation
              logoUrl="/images/alcumus-logo-with-tagline.svg"
              logoAltText="Alcumus logo"
              logoRedirect="/"
              showDrawer
              userContent={<UserMenu user={user} onLogout={handleLogout} />}
              drawerContent={
                drawerItems.length && (
                  <>
                    {showBackButton && <BackButton />}
                    <DrawerContent drawerItems={drawerItems} />
                  </>
                )
              }
            />
            <main className={classes.content}>
              <MToolbar />
              {children}
            </main>
          </div>
        </ToastContextProvider>
      </PageContext.Provider>
    </Page>
  );
}
