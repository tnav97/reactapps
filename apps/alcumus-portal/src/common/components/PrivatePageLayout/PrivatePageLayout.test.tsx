import React from 'react';
import { render } from '@testing-library/react';
import store from 'store';
import { MemoryRouter } from 'react-router-dom';
import PrivatePageLayout from './PrivatePageLayout';
import { defaultTestInitialState } from '../../../../tests/defaultTestInitialState';
import StateDecorator from '../../../client/redux/StateDecorator';

describe('PrivatePageLayout', () => {
  beforeAll(() => {
    store.set('refreshToken', 'aReallyLongRefreshToken');
    store.set('expiresIn', 300);
  });

  afterAll(() => {
    store.remove('refreshToken');
    store.remove('expiresIn');
  });

  test('it should match snapshot with user profile not having name set', () => {
    const { container } = render(
      <MemoryRouter>
        <StateDecorator initialState={defaultTestInitialState}>
          <PrivatePageLayout
            user={{ email: 'alcumusdeveloper@gmail.com' }}
            fetchUserProfile={() => {}}
            refreshTokens={() => {}}
          >
            <h1>Test Page</h1>
          </PrivatePageLayout>
        </StateDecorator>
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });

  test('it should match snapshot with user profile having name set', () => {
    const { container } = render(
      <MemoryRouter>
        <StateDecorator initialState={defaultTestInitialState}>
          <PrivatePageLayout
            user={{
              firstName: 'foo',
              lastName: 'bar',
              email: 'alcumusdeveloper@gmail.com',
            }}
            fetchUserProfile={() => {}}
            refreshTokens={() => {}}
          >
            <h1>Test Page</h1>
          </PrivatePageLayout>
        </StateDecorator>
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });
});
