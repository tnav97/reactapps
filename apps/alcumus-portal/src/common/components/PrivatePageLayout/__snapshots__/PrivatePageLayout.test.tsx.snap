// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivatePageLayout it should match snapshot with user profile having name set 1`] = `
<div>
  <div
    class="MuiContainer-root makeStyles-page-31 makeStyles-portalPage-26 makeStyles-portalPage-30"
  >
    <div
      class="makeStyles-root-24"
    >
      <header
        class="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary makeStyles-appBar-32 mui-fixed MuiPaper-outlined"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
        >
          <div
            class="makeStyles-appBarContent-33"
          >
            <button
              aria-label="side menu"
              class="MuiButtonBase-root MuiIconButton-root makeStyles-sideMenuButton-37 MuiIconButton-colorInherit"
              data-testid="sideMenuIcon"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <span
                  class="material-icons"
                  data-testid="icon-close"
                >
                  close
                </span>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <div
              class="makeStyles-logo-34"
            >
              <a
                aria-current="page"
                class="active"
                href="/"
              >
                <img
                  alt="Alcumus logo"
                  data-testid="logoImage"
                  src="/images/alcumus-logo-with-tagline.svg"
                />
              </a>
            </div>
            <div
              class="makeStyles-userContent-35"
            >
              <div
                class="makeStyles-desktopUserMenu-45"
              >
                <div
                  class="makeStyles-userProfileMenu-41"
                >
                  <img
                    alt="User Profile"
                    aria-controls="primary-search-account-menu"
                    aria-haspopup="true"
                    aria-label="account of current user"
                    class="makeStyles-userProfile-42"
                    data-testid="profileImage"
                    src="/icons/Profile.svg"
                  />
                </div>
              </div>
              <div
                class="makeStyles-mobileUserMenu-46"
              >
                <button
                  aria-label="Settings"
                  class="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-40 MuiIconButton-colorInherit"
                  data-testid="settingsIcon"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <span
                      class="material-icons-outlined"
                      data-testid="icon-settings"
                    >
                      settings
                    </span>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </header>
      <main
        class="makeStyles-content-25"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
        />
        <h1>
          Test Page
        </h1>
      </main>
    </div>
  </div>
</div>
`;

exports[`PrivatePageLayout it should match snapshot with user profile not having name set 1`] = `
<div>
  <div
    class="MuiContainer-root makeStyles-page-8 makeStyles-portalPage-3 makeStyles-portalPage-7"
  >
    <div
      class="makeStyles-root-1"
    >
      <header
        class="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary makeStyles-appBar-9 mui-fixed MuiPaper-outlined"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
        >
          <div
            class="makeStyles-appBarContent-10"
          >
            <button
              aria-label="side menu"
              class="MuiButtonBase-root MuiIconButton-root makeStyles-sideMenuButton-14 MuiIconButton-colorInherit"
              data-testid="sideMenuIcon"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <span
                  class="material-icons"
                  data-testid="icon-close"
                >
                  close
                </span>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <div
              class="makeStyles-logo-11"
            >
              <a
                aria-current="page"
                class="active"
                href="/"
              >
                <img
                  alt="Alcumus logo"
                  data-testid="logoImage"
                  src="/images/alcumus-logo-with-tagline.svg"
                />
              </a>
            </div>
            <div
              class="makeStyles-userContent-12"
            >
              <div
                class="makeStyles-desktopUserMenu-22"
              >
                <div
                  class="makeStyles-userProfileMenu-18"
                >
                  <img
                    alt="User Profile"
                    aria-controls="primary-search-account-menu"
                    aria-haspopup="true"
                    aria-label="account of current user"
                    class="makeStyles-userProfile-19"
                    data-testid="profileImage"
                    src="/icons/Profile.svg"
                  />
                </div>
              </div>
              <div
                class="makeStyles-mobileUserMenu-23"
              >
                <button
                  aria-label="Settings"
                  class="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-17 MuiIconButton-colorInherit"
                  data-testid="settingsIcon"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <span
                      class="material-icons-outlined"
                      data-testid="icon-settings"
                    >
                      settings
                    </span>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </header>
      <main
        class="makeStyles-content-2"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
        />
        <h1>
          Test Page
        </h1>
      </main>
    </div>
  </div>
</div>
`;
